### PR TITLE
fix(trading): show zero network fee for gasless 1inch Fusion quotes

### LINKED ISSUE #27911: Fix network fees for 1inch fusion
Show 0 ETH and $0 for network fees in trade form and trade preview when 1INCH FUSION is selected as provider.

Image

### PR DESCRIPTION
**Pull request by sherpaPSX**
## Description

- detect gasless exchange quotes via a shared trading helper based on EIP-712 sign data
- show zero network fee in the Exchange form Maximum fee row for 1inch Fusion offers
- show zero network fee in selected offer details and trade detail for gasless exchange quotes
- keep non-gasless exchange quotes on the existing composed transaction fee path
- add unit coverage for gasless quote detection and displayed network fee mapping

## Notes for QA

- Open Wallet > Trading > Exchange, select a 1inch Fusion offer, and verify the Maximum fee row shows 0 ETH and $0.
- Continue to selected offer details and trade detail for the same Fusion flow, and verify network fee stays 0; switch to a non-gasless quote and verify the regular network fee is still shown.
- I tested this manually, but please confirm that you can perform a swap on an account with 0 native tokens and that the form validation allows you to do so (you must have approval before, you know ;) ).

## Related Issue

Resolve #27911

## Screenshots:

image

## 🔍 Currents Test Results

🔍 **Suite native android test results:** [View in Currents](https://app.currents.dev/projects/iUe1Y4/runs?branch=feat/trading-27911-fusion-fee&lookback=7)

🔍 **Suite desktop test results:** [View in Currents](https://app.currents.dev/projects/4ytF0E/runs?branch=feat/trading-27911-fusion-fee&lookback=7)

🔍 **Suite web test results:** [View in Currents](https://app.currents.dev/projects/Og0NOQ/runs?branch=feat/trading-27911-fusion-fee&lookback=7)

## 🔒 Quarantined E2E Tests

<details><summary>Trezor Suite (web) — 6 test(s)</summary>

| Test | Type |
|------|------|
| Account metadata > dropbox provider | 🤖 auto |
| Remembered wallet loading > Load remembered wallet and open receive forms without connected device | 🤖 auto |
| Quarantine test: "Onboarding - create wallet,Success (basic)" | 🙋 manual |
| Quarantine test: "Database migration,Db migration between: release/22.5/web => develop/web" | 🙋 manual |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |
| Quarantine test: "Recovery T2T1 - dry run,Recovery after partial recovery" | 🙋 manual |

</details>

_Updated: 2026-06-16T07:42:25.167Z • 6 test(s) total_

<details><summary>Trezor Suite (desktop) — 3 test(s)</summary>

| Test | Type |
|------|------|
| Send Base > User can perform ethereum sending on base network | 🤖 auto |
| Labeling migration > Migration from local file | 🤖 auto |
| Quarantine test: "Multiple sessions,Session overtaken by another" | 🙋 manual |

</details>

_Updated: 2026-06-16T07:41:41.254Z • 3 test(s) total_

## 🌐 Preview deployments

🌐 **Suite Web preview:** [https://dev.suite.sldev.cz/suite-web/feat/trading-27911-fusion-fee/web/](https://dev.suite.sldev.cz/suite-web/feat/trading-27911-fusion-fee/web/)

<details>
<summary><h3>🤖 LLM Test Recommendations</h3></summary>

**Summary:** This change set modifies swap/exchange-specific UI components (form inputs, offer details, detail view) along with shared exchange utility functions and trading selectors. The core risk is in the swap/exchange flow — form rendering, offer confirmation display, and transaction detail views. The two files mapping to 121 tests (tradingSelectors.ts, exchangeUtils.ts) are broad shared utilities, but the changes are exchange-specific, so only swap/exchange tests need to run. The recommended strategy prioritizes the mocked swap E2E tests that directly exercise all three view components, with medium priority for fee-specific and input-validation swap tests, and low priority for the slower live swap tests.

<details><summary><b>Changed files (6)</b></summary>

- `packages/suite/src/views/wallet/trading/common/TradingDetail/TradingDetailExchange/TradingDetailExchange.tsx`
- `packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx`
- `packages/suite/src/views/wallet/trading/common/TradingSelectedOffer/TradingOfferExchange/TradingOfferExchangeDetails.tsx`
- `suite-common/trading/src/selectors/tradingSelectors.ts`
- `suite-common/trading/src/utils/exchange/__tests__/exchangeUtils.test.ts`
- `suite-common/trading/src/utils/exchange/exchangeUtils.ts`

</details>

**Recommended tests (12)**

<details><summary>🔴 <b>High priority (4)</b></summary>

- `suite/e2e/tests/trading/swap-coin-to-token.test.ts` — Directly exercises the swap flow end-to-end including TradingDetailExchange (swap detail status/success), TradingExchangeFormInputs (swap form fill), and TradingOfferExchangeDetails (confirmation with exchange type, provider, amounts). All three view-layer changed files are exercised.
- `suite/e2e/tests/trading/swap-coins.test.ts` — Full swap flow (SOL→BTC) that validates swap form inputs, offer confirmation details (exchange type, provider name), transaction status progression through multiple states, and the detail exchange view. Covers all three swap-specific changed view files plus exercises exchangeUtils through status/detail rendering.
- `suite/e2e/tests/trading/swap-token-to-coin.test.ts` — Tests token-to-coin swap which exercises the exchange form inputs (asset selection with token), offer exchange details (confirmation amounts, exchange type, provider), and detail exchange (success status). Covers a different asset combination than swap-coins, increasing coverage breadth.
- `suite/e2e/tests/trading/swap-tokens.test.ts` — Tests token-to-token swap (USDT→USDC on Solana) exercising all three changed view files. Validates exchange form inputs with token assets on both sides, offer exchange details confirmation, and detail exchange success status.

</details>

<details><summary>🟡 <b>Medium priority (6)</b></summary>

- `suite/e2e/tests/trading/swap-fees-bitcoin.test.ts` — Tests the swap form with custom Bitcoin fee settings, exercising TradingExchangeFormInputs (form filling) and TradingOfferExchangeDetails (confirmation modal with fee details). Validates fee calculation path that may be affected by exchangeUtils changes.
- `suite/e2e/tests/trading/swap-fees-ethereum.test.ts` — Tests swap form with custom Ethereum gas fees, exercising TradingExchangeFormInputs and TradingOfferExchangeDetails. Validates Ethereum-specific fee parameters that flow through the exchange utils and confirmation components.
- `suite/e2e/tests/trading/swap-history.test.ts` — Tests swap trade history display including transaction detail sidebar with send/receive amounts, provider, status, and order ID. The TradingDetailExchange component renders swap detail views, and tradingSelectors likely provides data for the history/detail views.
- `suite/e2e/tests/trading/swap-inputs.test.ts` — Directly tests the swap form input interactions including asset selection, amount entry, fraction buttons, provider selection, and validation errors. TradingExchangeFormInputs is the primary component under test, and tradingSelectors feeds data to it.
- `suite/e2e/tests/trading/swap-token-to-disabled-network.test.ts` — Tests swap form behavior when buy asset is on a disabled network, directly exercising TradingExchangeFormInputs for asset selection and quote display. Validates that the form correctly handles edge-case network configurations.
- `suite/e2e/tests/trading/navigation.test.ts` — Tests navigating to the Swap form from multiple entry points and verifying it opens with the correct pre-selected cryptocurrency. TradingExchangeFormInputs renders the swap form that is verified after navigation.

</details>

<details><summary>🟢 <b>Low priority (2)</b></summary>

- `suite/e2e/tests/trading-live/live-swap-spl-token-to-coin.test.ts` — Live swap test that exercises the full swap flow including form inputs, detail exchange, and offer exchange details. Lower priority because live tests are slower and the mocked swap tests already cover the changed components, but this validates against real exchange behavior.
- `suite/e2e/tests/trading-live/live-swap-coin-to-token.test.ts` — Live swap test covering SOL→USDC flow. All three view-layer changed files are exercised but this duplicates coverage already provided by the mocked swap tests at higher priority. Useful only for live environment validation.

</details>

⚠️ **Changes with no test coverage (1)**
- `suite-common/trading/src/utils/exchange/__tests__/exchangeUtils.test.ts`

_Updated: 2026-06-16T07:38:38.367Z_

</details>

### CODE REVIEW

**Comment by sherpaPSX** [comment]
@coderabbit review

File: suite-common/trading/src/selectors/tradingSelectors.ts [hunk 1]
```diff
@@ -632,6 +634,12 @@ export const selectTradingSellFormStep = (state: TradingRootState) =>
 export const selectTradingComposedTransactionInfo = (state: TradingRootState) =>
     state.wallet.trading.composedTransactionInfo;
 
+export const selectTradingDisplayComposedFee = (
```
Comment (vytick) [inline suite-common/trading/src/selectors/tradingSelectors.ts:636]: we probably should memoize this

File: suite-common/trading/src/utils/exchange/exchangeUtils.ts [hunk 2]
```diff
@@ -122,6 +126,27 @@ export const requiresTokenApproval = (quote?: ExchangeTrade): boolean =>
...
+    quote: ExchangeTrade | undefined,
+    fee: string | undefined,
+): string | undefined => (hasEip712SignDataType(quote) ? '0' : fee);
+
+export const getDisplayComposedLevels = <T extends GeneralPrecomposedLevels>(
+    quote: ExchangeTrade | undefined,
+    composedLevels: T | undefined,
+): T | undefined => {
+    if (!composedLevels || !hasEip712SignDataType(quote)) {
+        return composedLevels;
```
Comment (vytick) [inline suite-common/trading/src/utils/exchange/exchangeUtils.ts]: __Q:__ it seems we rely on this returning undefined. 

a) But it is a bit cryptic, can we return undefined directly? 
b) 
And actually if we do rely on this, in case composedLevels it is defined && it is eip712, it returns actual defined levels. Or do I get it wrong? 

In case we want this, can we have a test case for this?

File: packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx [hunk 3]
```diff
@@ -56,8 +57,14 @@ export const TradingExchangeFormInputs = () => {
         shouldSendInSats,
         showReserveBanner,
         resetSelectedOffer,
+        selectedQuote,
         setAmountLimits,
     } = context;
+
+    const displayComposedLevels = useMemo(
```
Comment (vytick) [inline packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx:64]: can we make this into memoized selector, pls?

File: suite-common/trading/src/utils/exchange/exchangeUtils.ts (hunk 2 above)
Comment (sherpaPSX) [reply suite-common/trading/src/utils/exchange/exchangeUtils.ts]: a) No, we can't. This only applies to non-EIP712 cases, so `composedLevels` are passed through unchanged.
b) I'll rewrite the condition in a positive form to improve readability ;)

test for `undefined` state added

File: suite-common/trading/src/selectors/tradingSelectors.ts (hunk 1 above)
Comment (sherpaPSX) [reply suite-common/trading/src/selectors/tradingSelectors.ts:636]: There is any computation and it returns primitive. IMO zero win. Am I wrong?

File: packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx (hunk 3 above)
Comment (sherpaPSX) [reply packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx:64]: Values are from the context, not from the Redux. IMO memo is correct here ;)

Comment (vytick) [reply packages/suite/src/views/wallet/trading/common/TradingForm/TradingExchangeFormInputs.tsx:64]: Makes sense in that case, I believed that  both composedLevels and selectedQuote are in the redux

**Review by vytick — APPROVED** [review]
lgtm
